toolchain/check/eval.cpp

```diff
@@ -17,6 +17,104 @@
+    // them.
+    if (specific_eval_info) {
+      const auto& symbolic_info =
+          context.constant_values().GetSymbolicConstant(const_id);
+      if (symbolic_info.index.is_valid() &&
+          symbolic_info.generic_id ==
+              context.generic_instances().Get(specific_id).generic_id &&
+          symbolic_info.index.region() == specific_eval_info->region) {
+        auto inst_id = specific_eval_info->values[symbolic_info.index.index()];
+        CARBON_CHECK(inst_id.is_valid())
+            << "Forward reference in eval block: index "
+            << symbolic_info.index.index() << " referenced before evaluation";
+        return context.constant_values().Get(inst_id);
+      }
+    }
+
+    // Map from a specific constant value to the canonical value.
+    return GetConstantInInstance(context.sem_ir(), specific_id, const_id);
+  }
+
+  // Gets the constant value of the specified instruction in this context.
+  auto GetConstantValue(SemIR::InstId inst_id) -> SemIR::ConstantId {
+    return GetInContext(context.constant_values().Get(inst_id));
+  }
+
+  // Gets the constant value of the specified type in this context.
+  auto GetConstantValue(SemIR::TypeId type_id) -> SemIR::ConstantId {
+    return GetInContext(context.types().GetConstantId(type_id));
+  }
+
+  // Gets the instruction describing the constant value of the specified type in
+  // this context.
+  template <typename IdT>
+  auto GetConstantValueAsInst(IdT id) -> SemIR::Inst {
+    return insts().Get(
+        context.constant_values().GetInstId(GetConstantValue(id)));
+  }
+
+  auto ints() -> CanonicalValueStore<IntId>& { return sem_ir().ints(); }
+  auto floats() -> FloatValueStore& { return sem_ir().floats(); }
+  auto bind_names() -> SemIR::BindNameStore& { return sem_ir().bind_names(); }
+  auto functions() -> const ValueStore<SemIR::FunctionId>& {
+    return sem_ir().functions();
+  }
+  auto classes() -> const ValueStore<SemIR::ClassId>& {
+    return sem_ir().classes();
+  }
+  auto interfaces() -> const ValueStore<SemIR::InterfaceId>& {
+    return sem_ir().interfaces();
+  }
+  auto generic_instances() -> const SemIR::GenericInstanceStore& {
+    return sem_ir().generic_instances();
+  }
+  auto types() -> const SemIR::TypeStore& { return sem_ir().types(); }
+  auto type_blocks() -> SemIR::BlockValueStore<SemIR::TypeBlockId>& {
+    return sem_ir().type_blocks();
+  }
+  auto insts() -> const SemIR::InstStore& { return sem_ir().insts(); }
+  auto inst_blocks() -> SemIR::InstBlockStore& {
+    return sem_ir().inst_blocks();
+  }
+
+  auto sem_ir() -> SemIR::File& { return file; }
+
+  auto emitter() -> Context::DiagnosticEmitter& { return context.emitter(); }
+
+  // The type-checking context in which we're performing evaluation.
+  Context& context;
+  // A cached reference to the file, to avoid a double indirection when
+  // accessing its value stores.
+  SemIR::File& file = context.sem_ir();
```

Comment:
To note, this feels odd to me: it's unusual, and seems like it'd have a negligible effect on performance. Maybe I'm thinking about that wrong though?

The flipside is that all the internal `sem_ir()` calls look odd; if you keep this, maybe change them to reference the cached reference directly?

Reply:
Removed for now; we can re-add if we have evidence of this being an issue.